Add basic documentation for beginners

mbuesch opened the pull request:
There are many different traits and structs that interoperate in very complex ways in this crate. That makes it very hard to understand the basic concept of the crate. It's hard to find out where to look first in the documentation.
As always, examples help a lot when explaining things.

This pull request adds a couple of basic examples to help kick starting the knowledge of people new to this project.

Rahix commented on src/lib.rs:
````diff
@@ -63,6 +63,59 @@
+//!
+//! # How to use this crate?
+//!
+//! ## Main program entry point
+//!
+//! ```
+//! #[avr_device::entry]
+//! fn main() -> ! {
+//!     loop {
+//!         // Your code here.
+//!     }
+//! }
+//! ```
````
I think we also need to mention here that the `rt` crate feature must be active for this.

Rahix commented on src/interrupt.rs:
```diff
@@ -3,9 +3,27 @@
+//! use avr_device::interrupt::Mutex;
+//! use core::cell::Cell;
+//!
+//! static MYGLOBAL: Mutex<Cell<u16>> = Mutex::new(Cell::new(0));
```
I think it would be valuable to point out that `Cell` should be used for `Copy` types and `RefCell` when a reference to the data inside the mutex is needed at some point.

Rahix commented on src/lib.rs:
```diff
@@ -63,6 +63,59 @@
 //! * `attiny861`
 //! * `attiny88`
+//!
+//! # How to use this crate?
```
Maybe mention that this is the low-level peripheral access crate (PAC) and there are higher level crates like `avr-hal` which implement a much easier to use interface ontop.

Rahix reviewed (commented):
Hey, thank you very much for the PR, I really like this idea and I think we should actually document a lot more than just that.  To start, I've left a few comments below.

mbuesch commented:
Thanks for your comments.
I implemented all your suggestions. And I also did the following changes:

* Collapsed the supported device list. Now it takes less valuable first-sight screen space. But it's a bit harder to find a specific device in the list. What do you think? I think it's worth the trade off.
* Added a small section talking about crate features.

Rahix commented on src/interrupt.rs:
```diff
@@ -3,9 +3,29 @@
+//! use core::cell::Cell;
+//!
+//! // Use Cell, if the wrapped type is Copy.
+//! // Use RefCell, if the wrapped type is not Copy or if you need a mut reference to it for other reasons.
```
Actually, you will also need the RefCell to get an immutable reference...

```suggestion
//! // Use RefCell, if the wrapped type is not Copy or if you need a reference to it for other reasons.
```

mbuesch replied:
I think we don't need a (Ref)Cell at all, if only an immutable ref is needed. Mutex::borrow() returns an immutable ref.
In fact, we wouldn't even need a Mutex for immutable-only global access.
Or am I missing something?

Rahix replied:
Right, but at some point you'll want to fill the Mutex so you need the mutable access at least once.  I don't think there is way around a RefCell then, even if you only grab immutable references later...

mbuesch replied:
Ok. It could be static initialization. But this is theoretical anyway. Almost all uses would need Cell or RefCell mutable. Therefore, I removed the word.

Rahix commented:
> Collapsed the supported device list. Now it takes less valuable first-sight screen space. But it's a bit harder to find a specific device in the list. What do you think? I think it's worth the trade off.

Agreed, I also prefer it like this.  If people need to find a device, they can `CTRL-F`...

> Added a small section talking about crate features.

Also a good idea!  Thanks!

Rahix reviewed (approved):
Perfect, thanks a lot!
